pghoard/rohmu/object_storage/s3.py

```diff
@@ -62,152 +77,245 @@ def get_metadata_for_key(self, key):
         return self._metadata_for_key(key)
 
     def _metadata_for_key(self, key):
-        item = self.bucket.get_key(key)
-        if item is None:
-            raise FileNotFoundFromStorageError(key)
+        try:
+            response = self.s3_client.head_object(Bucket=self.bucket_name, Key=key)
+        except botocore.exceptions.ClientError as ex:
+            status_code = ex.response.get("ResponseMetadata", {}).get("HTTPStatusCode")
+            if status_code == 404:
+                raise FileNotFoundFromStorageError(key)
+            else:
+                raise StorageError("Metadata lookup failed for {}".format(key)) from ex
 
-        return item.metadata
+        return response["Metadata"]
 
     def delete_key(self, key):
         key = self.format_key_for_backend(key)
         self.log.debug("Deleting key: %r", key)
-        item = self.bucket.get_key(key)
-        if item is None:
-            raise FileNotFoundFromStorageError(key)
-        item.delete()
+        self._metadata_for_key(key)  # check that key exists
+        self.s3_client.delete_object(Bucket=self.bucket_name, Key=key)
 
     def list_iter(self, key, *, with_metadata=True):
         path = self.format_key_for_backend(key, trailing_slash=True)
         self.log.debug("Listing path %r", path)
-        for item in self.bucket.list(path, "/"):
-            if not hasattr(item, "last_modified"):
-                continue  # skip objects with no last_modified: not regular objects
-            name = self.format_key_from_backend(item.name)
-            if name == path:
-                continue  # skip the path itself
-            if with_metadata:
-                metadata = self._metadata_for_key(item.name)
+        continuation_token = None
+        while True:
+            if continuation_token:
+                extra_args = {
+                    "ContinuationToken": continuation_token,
+                }
+            else:
+                extra_args = {}
+            response = self.s3_client.list_objects_v2(
+                Bucket=self.bucket_name,
+                Delimiter="/",
+                Prefix=path,
+                **extra_args,
+            )
+            for item in response.get("Contents", []):
+                if with_metadata:
+                    metadata = self._metadata_for_key(item["Key"])
```

Comment:
I think `_metadata_for_key` expects a formatted key - how did this pass tests?  Nothing calls this with `with_metadata` set to True?

Reply:
Item["Key"] is coming from the list_objects response and does thus carry the correct, actual and full S3 path. It was never passed back out to the caller and thus didn't go through format_key_from_backend().

Reply:
Right -- a few lines below you use format_key_**FROM**_backend, I thought you had the key formatting in a wrong place.